src/scales/scale.time.js

```diff
@@ -334,27 +334,42 @@ function generate(min, max, minor, major, capacity, options) {
  * Returns the right and left offsets from edges in the form of {left, right}.
  * Offsets are added when the `offset` option is true.
  */
-function computeOffsets(table, ticks, min, max, options) {
+function computeOffsets(table, ts, min, max, options) {
 	var left = 0;
 	var right = 0;
-	var upper, lower;
-
-	if (options.offset && ticks.length) {
-		if (!options.time.min) {
-			upper = ticks.length > 1 ? ticks[1] : max;
-			lower = ticks[0];
-			left = (
-				interpolate(table, 'time', upper, 'pos') -
-				interpolate(table, 'time', lower, 'pos')
-			) / 2;
-		}
-		if (!options.time.max) {
-			upper = ticks[ticks.length - 1];
-			lower = ticks.length > 1 ? ticks[ticks.length - 2] : min;
-			right = (
-				interpolate(table, 'time', upper, 'pos') -
-				interpolate(table, 'time', lower, 'pos')
-			) / 2;
+	var timestamps = [];
+	var i, ilen, timestamp, length, upper, lower, divisor;
+
+	if (options.offset) {
+		// First, it processes `ts.labels` to pick up timestamps of labels, which are
+		// primarily used for offset calculation because bars are usually on labels.
+		// But if there are no labels, it should add offsets based on `ts.data` because
+		// we support bars with {x, y} data points as well. That is the second path.
+		[ts.labels, ts.data].forEach(function(data) {
```

Comment:
Would `ts.data` be enough since most of the time [it contains `ts.labels`](/chartjs/Chart.js/blob/master/src/scales/scale.time.js#L506)?

Reply:
@simonbrunel I was considering the case of a mixed chart with bars on labels and line with {x, y} points. If the label positions and {x, y} points are not aligned on a timeline, bar width calculation could go wrong. But, as you say, that may be an edge case, and I agree that the calculation based on `ts.data` works in most cases.